### ISSUE TITLE
Detect intended out argument on mutability and lifetime errors involving fn args

### ISSUE THREAD

**Issue by estebank** [rust-lang/rust#81976]
Given code like:

```rust
fn foo(x: &String) {
    let y = "asdf".to_string();
    x = &y;
}
```

we [currently emit](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=dd257a6e7529f899a57540fe9fec3462):

```
error[E0384]: cannot assign to immutable argument `x`
 --> src/main.rs:3:5
  |
1 | fn foo(x: &String) {
  |        - help: make this binding mutable: `mut x`
2 |     let y = "asdf".to_string();
3 |     x = &y;
  |     ^^^^^^ cannot assign to immutable argument

error[E0597]: `y` does not live long enough
 --> src/main.rs:3:9
  |
1 | fn foo(x: &String) {
  |           - let's call the lifetime of this reference `'1`
2 |     let y = "asdf".to_string();
3 |     x = &y;
  |     ----^^
  |     |   |
  |     |   borrowed value does not live long enough
  |     assignment requires that `y` is borrowed for `'1`
4 | }
  | - `y` dropped here while still borrowed
```

We should detect cases like these and instead [suggest making `x: &mut String`](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=94c05485c82d5668d6edc7b422fcd646) which is the more likely desire:

```rust
fn foo(x: &mut String) {
    let y = "asdf".to_string();
    *x = y;
}
```

**Comment by estebank** [comment]
_Sourced from https://twitter.com/SpikePuppet/status/1357242364520144898_
